### DISCUSSION TITLE
DDN Exascaler with slurm v6 modules

### DISCUSSION THREAD

**Discussion by tdido** [GoogleCloudPlatform/cluster-toolkit#2223] [Q&A] [answered]
I was wondering whether it should be possible to use the exascaler modules with slurm v6. As it is, if I include such filesystem in my blueprint Slurm is not able to start. Before delving deeper into the issues I though I'd check whether it's compatible at all.

**Comment by mr0re1** [comment]
@tdido , could you please provide more details (share blueprint) ? 
I've just run a simple test of V6 with `community/modules/file-system/DDN-EXAScaler`, it WAI.

**Reply by tdido** [reply]
Sure thing, here you go. Thanks!

[cluster.yaml.gz](https://github.com/GoogleCloudPlatform/hpc-toolkit/files/14383062/cluster.yaml.gz)

**Reply by mr0re1** [reply]
Thank you for sharing the blueprint. It looks good at the glance. I din't try to reproduce it yet, due to usage of `pre-existing-vpc`.
Could you please also elaborate on "Slurm is not able to start", how does it manifest? any deployment errors, slurm logs ?

If you will do try deployment again, please use `develop` branch. Thank you.

**Reply by tdido** [reply]
Thank you again for looking into this, @mr0re1.

I'm attaching a few log files: in `error.txt`you can see the error upon trying `squeue`. I also include Slurm's `setup.log`, and the entire `/var/log` contents from the login node.

[logs.tar.gz](https://github.com/GoogleCloudPlatform/hpc-toolkit/files/14392245/logs.tar.gz)

The most noticeable error is a failure upon the automatic mounting the controller's `/opt/apps` on the login node. As I mentioned, this does not occur if I don't include the exascaler module and leave everything else the same.

**Reply by tdido** [reply]
Quick update: I tried the same deployment with the latest commit from `develop`. `ghpc create` refused to run saying that there was an image incompatibility. It offered some options and I updated to `slurm-gcp-6-4-hpc-rocky-linux-8`. It then create the cluster but the same error remains.

**Reply by mr0re1** [reply]
@tdido  , thank you for sharing ! Looks like the problem is in VPC setup, my guess it has nothing to do with DDN.
```
squeue: error: resolve_ctls_from_dns_srv: res_nsearch error: Unknown host
squeue: error: fetch_config: DNS SRV lookup failed
```

Did you try to create minimal cluster (no NFSes, just a Slurm controller + single login node)  on `cluster1-vpc-01` ?

**Reply by mr0re1** [reply]
> refused to run saying that there was an image incompatibility

That's expected, you could simply remove any mentions of `instance_image` from your blueprint, Slurm modules will default to an appropriate Rocky8 image.

**Reply by tdido** [reply]
Thanks, @mr0re1.

I just tried to spin up the cluster by just removing the `- scratchfs` entries from the `use` sections in `main_partition` and `slurm_controller`. After this, I can spin up the cluster and it works fine (the DDN VMs get created but are not mounted).

So it would seem that if it is indeed a VPC setup problem it's specific to the login and/or controller talking to the DDN machines.

**Reply by tdido** [reply]
To rule out the pre-existing VPC problem I've switched `network1` to `modules/network/vpc`, but I'm still getting the same errors.

**Comment by mr0re1** [comment]
> I've switched network1 to modules/network/vpc, but I'm still getting the same errors.

Thank you for doing that, I will proceed with trying your blueprint (with new `vpc`)

**Reply by mr0re1** [reply]
I couldn't reproduce it yet, DDN WAI:
```sh
@login $ srun -N 1 -p n2 cat /storage/scratch01/hi
Hi
```
I will do a few more deployments test, will keep you in touch.

**Reply by mr0re1** [reply]
I was able to reproduce it. Strangely, specifying `instance_image` leads to failure to mount both filestore and exascaler.
I will continue investigating it, meanwhile could you please try simply removing any mentions of `instance_image` from your blueprint, hope it would help.

**Comment by mr0re1** [comment, marked as answer]
@tdido , I found the culprit. The `instance_image` declared as a "deployment variable" got propagated into `community/modules/file-system/DDN-EXAScaler/variables`. Your blueprint creates an EXAScaler instances with Slurm image, rendering them invalid.

Please remove `instance_image` from your blueprint deployment `vars`, or rename it to something else.
I'm closing this bug, please feel free to re-open if fix doesn't work for you or other problems occur.

**Reply by tdido** [reply]
Yes, that absolutely did it. Thank you very much for your support.

Would it be worth it to put some kind of validator in place to check for this kind of thing? Is it something that should be changed on the EXAScaler module? Happy to look into creating a PR.

**Reply by mr0re1** [reply]
> Would it be worth it to put some kind of validator in place to check for this kind of thing?
> Is it something that should be changed on the EXAScaler module? 

We are definitely considering it. It may require some lengthy discussions though.

> Happy to look into creating a PR.

@tdido thank you for offering help! If you want, you could update all examples to stop using global `var.instance_image`. That is a "no-brainer", I'm certain there will be no push-back on such contribution )
